src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -552,6 +582,67 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
         function_node.getArguments().getNodes() = std::move(or_operands);
         function_node.resolveAsFunction(or_function_resolver);
     }
+
+    void tryOptimizeOutRedundantEquals(QueryTreeNodePtr & node)
+    {
+        auto & function_node = node->as<FunctionNode &>();
+        assert(function_node.getFunctionName() == "equals");
+
+        bool lhs_const;
```

Comment:
nitpick: I'd rewrite code below a bit, to declare variables closer to first usages, like this (I haven't checked the code below, so feel free to change it or just discard suggestion):

```cpp
        const auto function_arguments = function_node.getArguments().getNodes();
        if (function_arguments.size() != 2)
            return;

        const auto & lhs = function_arguments[0];
        const auto & rhs = function_arguments[1];

        UInt64 constant_value;
        bool is_lhs_const;
        if (const auto * lhs_constant = lhs->as<ConstantNode>())
        {
            if (!lhs_constant->getValue().tryGet<UInt64>(constant_value) || constant_value > 1)
                return;
            is_lhs_const = true;
        }
        else if (const auto * rhs_constant = rhs->as<ConstantNode>())
        {
            if (!rhs_constant->getValue().tryGet<UInt64>(constant_value) || constant_value > 1)
                return;
            is_lhs_const = false;
        }
        else
            return;

        bool need_invert = (constant_value == 0);

        const FunctionNode * child_function = is_lhs_const ? rhs->as<FunctionNode>() : lhs->as<FunctionNode>();

        if (!child_function || !isBooleanFunction(child_function->getFunctionName()))
            return;
```